src/ShellCheck/Formatter/TTY.hs

```diff
@@ -147,17 +148,49 @@ showFixedString color comments lineNum fileLines =
     let line = fileLines !! fromIntegral (lineNum - 1) in
     -- need to check overlaps
     case filter (hasApplicableFix lineNum) comments of
-        (first:_) -> do
+        [] -> return ()
+        -- all the fixes are single-line only, but there could be multiple
+        -- fixes for that single line. We can fold the fixes (which removes
+        -- overlaps), and apply it as a single fix with multiple replacements.
+        applicableComments -> do
+            let mergedFix = (fold . catMaybes . (map pcFix)) applicableComments
             -- in the spirit of error prone
             putStrLn $ color "message" "Did you mean: "
-            putStrLn $ unlines $ fixedString first fileLines
-        _ -> return ()
-
-fixedString :: PositionedComment -> [String] -> [String]
-fixedString comment fileLines =
-    case (pcFix comment) of
-    Nothing -> [""]
-    Just fix -> case (fixReplacements fix) of
+            putStrLn $ unlines $ fixedString (makeNonVirtual2 mergedFix fileLines) fileLines
+
+-- Realign replacements from a tabstop of 8 to 1
+makeNonVirtual2 f ls =
```

Comment:
This approach would definitely work. 

I was thinking though.. what if we had a data structure that would represent a string, but also keep track of the original string indices.

For example, in linear form, it could be a `[(Int, String)]` where the int is the length of the piece, while the String is the contents. A trivial way to turn a string into such a form is e.g. `[(5, "hello)]`, or `[(1, "h"), (1, "e"), (1, "l"), (1, "l"), (1, "o")]`.

If you e.g. wanted to add a quote at index 0 and 5, you could find the index by counting, and then inserting to get `[(0, "\""), (5, "hello"), (0, "\"")]` .

The benefit of this is that when another fix wants to apply something at a given index, it can similarly count without having to adjust for previous string modifications. The other benefit is that by having `[(8, "\t")]`, tabs are automatically accounted for (at the start of the line anyways).

Realistically it would probably be in a tree form instead of a list for speed, with the same principle of letting you insert at original string indices.

Do you think such a thing could work?

Reply:
Yea this looks interesting, how will it work with multiple lines, i.e. how will "hello\nworld" be represented?
how about deletions? What will deleting the first char from `(5, "hello")` look like?

Reply:
I think for multiple lines it would be simpler to treat it as a single line containing new lines, and do the replacement logic for a single line. (a made-up problematic example would be if the fix is spans 3 rows but the range to be replaced spans 2 rows:
```
[ [(1, "h"), (2, "e")], [(1, "l"),(3, "lo")] ]
  -> [ [(1, "h"), (2, "e")], [(1, "l"),(1, "l")], [(1, "o")] ]
```
then for the last row `[(1, "o")]`, you probably need an indicator that it shouldn't be counted when counting line numbers for the next fix, (e.g. if a fix is for line 3, it's really the new line 4). So perhaps this structure will need a similar index for row, like it currently has for columns).

This is a good idea, having something more robust is good. But I don't have the knowledge or experience to know what the more robust thing looks like :) I'm leaning towards my crappy way of doing it, then filing a feature request with your notes for future implementation.